### PR TITLE
Use OMEMO for offline MUC members

### PR DESCRIPTION
**Pull request by paulfariello**

### CODE REVIEW

**Comment by jubalh** [comment]
@optmzr @mdosch @kaffeekanne could you help testing? :)

**Comment by mdosch** [comment]
I received a segfault:

```
profanity -v
Profanity, version 0.7.1dev.feature/omemo-muc-offline.0f0df628
Copyright (C) 2012 - 2019 James Booth <boothj5web@gmail.com>.
Copyright (C) 2019 Michael Vetter <jubalh@iodoru.org>.
License GPLv3+: GNU GPL version 3 or later <https://www.gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Build information:
XMPP library: libstrophe
Desktop notification support: Enabled
OTR support: Enabled (libotr 4.1.1)
PGP support: Enabled (libgpgme 1.13.1-unknown)
OMEMO support: Enabled
C plugins: Enabled
Python plugins: Enabled (2.7.17)
GTK icons: Enabled
```
OS: Debian Bullseye (Testing)

```
Program received signal SIGSEGV, Segmentation fault.
__strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:65
65	../sysdeps/x86_64/multiarch/strlen-avx2.S: Datei oder Verzeichnis nicht gefunden.
(gdb) bt
#0  __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:65
#1  0x00007ffff6d6145f in __GI___strdup (s=0x0) at strdup.c:41
#2  0x000055555558c8d7 in muc_members_remove (room=0x5555586f04c0 "golang@conference.samwhited.com", jid=0x0)
    at src/xmpp/muc.c:902
#3  0x000055555558c945 in muc_members_update (room=0x5555586f04c0 "golang@conference.samwhited.com", jid=0x0, 
    affiliation=0x5555586f1720 "none") at src/xmpp/muc.c:910
#4  0x00005555555a9805 in sv_ev_muc_occupant_online (room=0x5555586f04c0 "golang@conference.samwhited.com", 
    nick=0x555558700cd0 "Alex P", jid=0x0, role=0x555558702f00 "visitor", affiliation=0x5555586f1720 "none", 
    actor=0x0, reason=0x0, show=0x5555586e38d0 "online", status=0x0) at src/event/server_events.c:1203
#5  0x000055555559cecf in _muc_user_occupant_handler (stanza=0x555558702eb0) at src/xmpp/presence.c:858
#6  0x000055555559cfff in _muc_user_handler (stanza=0x555558702eb0) at src/xmpp/presence.c:895
#7  0x000055555559bb2d in _presence_handler (conn=0x555555d0d900, stanza=0x555558702eb0, userdata=0x555555f0d480)
    at src/xmpp/presence.c:396
#8  0x00007ffff6ea5ebf in handler_fire_stanza (conn=conn@entry=0x555555d0d900, stanza=stanza@entry=0x555558702eb0)
    at src/handler.c:124
#9  0x00007ffff6ea2a6b in _handle_stream_stanza (stanza=0x555558702eb0, userdata=0x555555d0d900) at src/conn.c:1181
#10 0x00007ffff6eaeb0f in _end_element (userdata=0x555555f0bfa0, name=<optimized out>) at src/parser_expat.c:190
#11 0x00007ffff5d1d8aa in doContent (parser=parser@entry=0x555555ee1c70, startTagLevel=startTagLevel@entry=0, 
    enc=<optimized out>, s=<optimized out>, end=<optimized out>, nextPtr=0x555555ee1ca0, haveMore=1 '\001')
    at ../../src/lib/xmlparse.c:2845
#12 0x00007ffff5d1e57c in contentProcessor (parser=0x555555ee1c70, start=<optimized out>, end=<optimized out>, 
    endPtr=<optimized out>) at ../../src/lib/xmlparse.c:2444
#13 0x00007ffff5d20b1c in XML_ParseBuffer (isFinal=0, len=<optimized out>, parser=0x555555ee1c70)
    at ../../src/lib/xmlparse.c:1893
#14 XML_ParseBuffer (parser=0x555555ee1c70, len=<optimized out>, isFinal=0) at ../../src/lib/xmlparse.c:1863
#15 0x00007ffff6ea5794 in xmpp_run_once (ctx=0x555555f0d480, timeout=10) at src/event.c:263
#16 0x0000555555591a00 in connection_check_events () at src/xmpp/connection.c:119
#17 0x0000555555590f0a in session_process_events () at src/xmpp/session.c:256
#18 0x000055555558a74e in prof_run (log_level=0x5555556407b0 "INFO", account_name=0x0, config_file=0x0)
    at src/profanity.c:129
#19 0x0000555555618d79 in main (argc=1, argv=0x7fffffffe1c8) at src/main.c:176
```

**Comment by optmzr** [comment]
I also get a segfault during start up.

```
Profanity, version 0.7.1dev.feature/omemo-muc-offline.0f0df628
Copyright (C) 2012 - 2019 James Booth <boothj5web@gmail.com>.
Copyright (C) 2019 Michael Vetter <jubalh@iodoru.org>.
License GPLv3+: GNU GPL version 3 or later <https://www.gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Build information:
XMPP library: libstrophe
Desktop notification support: Disabled
OTR support: Enabled (libotr 4.1.1)
PGP support: Enabled (libgpgme 1.13.0)
OMEMO support: Enabled
C plugins: Enabled
Python plugins: Enabled (2.7.16)
GTK icons: Enabled
```

Stacktrace (on Gentoo):
```
#0  0x00007ffff6f21da1 in __strlen_avx2 () at ../sysdeps/x86_64/multiarch/strlen-avx2.S:65
#1  0x00007ffff6e4275e in __GI___strdup (s=0x0) at strdup.c:41
#2  0x000055555558ca08 in muc_members_remove ()
#3  0x00005555555a300b in sv_ev_muc_occupant_online ()
#4  0x0000555555598b39 in _presence_handler ()
#5  0x00007ffff6f95e81 in handler_fire_stanza (conn=conn@entry=0x555555d08a10, stanza=stanza@entry=0x5555566a46b0) at src/handler.c:124
#6  0x00007ffff6f929ba in _handle_stream_stanza (stanza=0x5555566a46b0, userdata=0x555555d08a10) at src/conn.c:1179
#7  0x00007ffff6f9ed7c in _end_element (userdata=0x555555d93e10, name=<optimized out>, prefix=<optimized out>, uri=<optimized out>)
    at src/parser_libxml2.c:183
#8  0x00007ffff5fc51fe in  () at /usr/lib64/libxml2.so.2
#9  0x00007ffff5fcb919 in  () at /usr/lib64/libxml2.so.2
#10 0x00007ffff5fcd2d0 in xmlParseChunk () at /usr/lib64/libxml2.so.2
#11 0x00007ffff6f9f24f in parser_feed
    (parser=<optimized out>, chunk=chunk@entry=0x7fffffffd1b0 "<presence to='optmzr@<redacted>' from='profanity@rooms.dismail.de/CcxWrk' id='<redacted>'><x xmlns='http://jabber.org/protocol/muc#user'><item role='participant' affiliation='none'/><"..., len=<optimized out>)
    at src/parser_libxml2.c:279
#12 0x00007ffff6f9576c in xmpp_run_once (ctx=0x555555e014a0, timeout=10) at src/event.c:263
#13 0x0000555555590a5f in connection_check_events ()
#14 0x000055555558aea2 in prof_run ()
#15 0x0000555555587296 in main ()
```

Thank you for working on this! :D

File: src/xmpp/muc.c [hunk 1]
```diff
@@ -858,6 +864,54 @@ muc_anonymity_type(const char *const room)
...
+        }
+    }
+}
+
+void
+muc_members_remove(const char *const room, const char *const jid)
+{
+    ChatRoom *chat_room = g_hash_table_lookup(rooms, room);
+    if (chat_room) {
+        g_hash_table_remove(chat_room->members, strdup(jid));
```
Comment (jubalh) [inline src/xmpp/muc.c]: How does g_hash_table_remove() behave if the jid is not in the list?
Also why strdup() ?

File: src/xmpp/muc.c [hunk 2]
```diff
@@ -858,6 +864,54 @@ muc_anonymity_type(const char *const room)
...
+        return NULL;
+    }
+}
+
+void
+muc_members_add(const char *const room, const char *const jid)
+{
+    ChatRoom *chat_room = g_hash_table_lookup(rooms, room);
+    if (chat_room) {
+        if (g_hash_table_insert(chat_room->members, strdup(jid), NULL)) {
```
Comment (jubalh) [inline src/xmpp/muc.c]: can the jid be NULL here?

File: src/event/server_events.c [hunk 3]
```diff
@@ -1199,6 +1199,10 @@ sv_ev_muc_occupant_online(const char *const room, const char *const nick, const
 
     gboolean updated = muc_roster_add(room, nick, jid, role, affiliation, show, status);
```
Comment (jubalh) [inline src/event/server_events.c]: I think sometimes we dont get a jid here jsut the nick of the user. So later on jid will be NULL and strdup() will segfault.

Comment (paulfariello) [reply src/event/server_events.c]: The API of `muc_members_*` expect a jid. I pushed a fix that ensure that the API is correctly called.

File: src/xmpp/muc.c (hunk 2 above)
Comment (paulfariello) [reply src/xmpp/muc.c]: It can't.

**Comment by jubalh** [comment]
Joining `cpp@conference.anoxinon.me` I get: `20/12/19 11:17:57 ! Error retrieving member list: Administrator privileges required` as a normal member.
I have just generally omemo support in but don't write OMEMO encrypted in that channel.

File: src/xmpp/muc.c (hunk 1 above)
Comment (paulfariello) [reply src/xmpp/muc.c]: We should remove this spurious strdup. Thanks.

**Comment by mdosch** [comment]
* I created a MUC and configured it to be non-public and non-anonymous in profanity
* I joined with the JID used with profanity with another device
* I joined with another JID on the same server
* I joined with another JID on another server
* I enabled OMEMO in profanity
     * I was told the MUC must be configured non-public and non-anonymous (what I did already)
     * I left the MUC and joined again and I could enable OMEMO
* I sent an OMEMO encrypted message
     * One participant was not yet trusted
     * Profanity did not mention that the message won't be encrypted for one participants key
     * After trusting the fingerprint all participants went offline
* I took all other devices except profanity offline
* I sent messages from profanity
* I took the other devices online again
* The messages could successfully be decrypted there

**Comment by paulfariello** [comment]
Fixes #1143

File: src/xmpp/iq.c [hunk 4]
```diff
@@ -2434,3 +2445,12 @@ _iq_free_affiliation_set(ProfPrivilegeSet *affiliation_set)
...
     }
 }
+
+static void
+_iq_free_affiliation_list(ProfAffiliationList *affiliation_list)
+{
+    if (affiliation_list) {
+        free(affiliation_list->affiliation);
+        free(affiliation_list);
+    }
```
Comment (jubalh) [inline src/xmpp/iq.c]: Should we set affiliation_list to NULL afterwards? Is it possible that we will get here again?

File: src/xmpp/iq.c [hunk 5]
```diff
@@ -93,6 +93,11 @@ typedef struct privilege_set_t {
     char *privilege;
 } ProfPrivilegeSet;
 
+typedef struct affiliation_list_t {
+    char *affiliation;
+    bool show;
```
Comment (jubalh) [inline src/xmpp/iq.c]: How about changing `show` to `handle_ui` or `show_ui_message` or similar? Something a bit more descriptive.

**Comment by jubalh** [comment]
@mdosch if you have time you could test again :)

**Pull request merged**